proposals/lip_add_256_bit_integer_types_to_lisk_codec.md

```diff
@@ -0,0 +1,214 @@
+
+<table>
+  <tr>
+   <td><strong>If data is of "dataType"</strong>
+   </td>
+   <td><strong>JavaScript validation</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>uint256
+   </td>
+   <td>
+
+```js
+typeof data === "bigint" &&
+0n <= data < 2n ** 256n;
+```
+
+   </td>
+  </tr>
+  <tr>
+   <td>int256
+   </td>
+   <td>
+
+```js
+typeof data === "bigint" &&
+-(2n ** 255n) <= data < 2n ** 255n;
+```
+
+   </td>
+  </tr>
+</table>
+
+### Encoding
+
+#### uint256
+
+Values of data type _uint256_ are encoded using big-endian encoding with a fixed length of 32 bytes. This sequence of 32 bytes is then encoded like a [value of bytes](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0027.md#encoding-strings-and-bytes). In particular, the key is of [wire type](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0027.md#keys) 2.
+
+#### int256
+
+Values of data type _int256_ are encoded by taking the big-endian encoding of their [two’s complement](https://en.wikipedia.org/wiki/Two%27s_complement) representation using a fixed length of 32 bytes. The two's complement representation means that non-negative values are encoded in their byte representation, and negative integers are encoded by taking the two’s complement of the magnitude. The two’s complement of the magnitude means that the 256-bit representation of the magnitude is flipped (bitwise NOT), and then 1 is added to the result. In particular, the most significant bit of the encoded value is one if and only if the integer is negative.
+
+As for _uint256_, the sequence of 32 bytes is then encoded like a [value of bytes](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0027.md#encoding-strings-and-bytes). In particular, the key is of [wire type](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0027.md#keys) 2.
+
+### Type Conversion to Protobuf
+
+When [converting a Lisk JSON schema to a .proto file](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0027.md#appendix-b-json-schema-to-protobuf), properties of data type _uint256_ or _int256_ in the Lisk JSON schema are converted to fields of type _bytes_ in the `.proto` file.
+
+## Rationale
+
+### Why Byte Representation Instead of Varint?
+
+Using varint encoding would break the compatibility with protobuf heavily. Recall that Lisk codec was designed in a way that [every Lisk JSON schema can be converted to a .proto file](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0027.md#appendix-b-json-schema-to-protobuf) and that every protobuf decoder decodes binary messages in the same way as Lisk codec using this `.proto` file. Using varint encoding would result in encoded values up to 37 bytes. There is, however, no field type in protobuf to which a protobuf decoder could decode a varint encoded value of more than 10 bytes. Hence, one could not convert a Lisk JSON schema that uses _uint256_ or _int256_ to a `.proto` file that could be used by a protobuf decoder for the same binary messages.
```

Comment:
i) I would explain why it is `37 bytes`.
ii) Also, I would add a reference of 10 bytes. For example, https://protobuf.dev/programming-guides/encoding/#varints or similar one.

Reply:
I addressed your points.